torch/distributed/rpc/api.py

```diff
@@ -207,6 +208,7 @@ def shutdown(graceful=True):
 
     if graceful:
         _wait_all_workers()
+        _delete_all_user_rrefs()
```

Comment:
I am trying to figure out why the test is flaky on this PR. 

```
Jan 17 02:11:26 Traceback (most recent call last):
Jan 17 02:11:26   File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
Jan 17 02:11:26     self.run()
Jan 17 02:11:26   File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
Jan 17 02:11:26     self._target(*self._args, **self._kwargs)
Jan 17 02:11:26   File "/var/lib/jenkins/workspace/test/common_distributed.py", line 198, in _run
Jan 17 02:11:26     getattr(self, test_name)()
Jan 17 02:11:26   File "/var/lib/jenkins/workspace/test/dist_utils.py", line 75, in new_test_method
Jan 17 02:11:26     rpc.shutdown(graceful=clean_shutdown)
Jan 17 02:11:26   File "/opt/conda/lib/python3.6/site-packages/torch/distributed/rpc/api.py", line 69, in wrapper
Jan 17 02:11:26     return func(*args, **kwargs)
Jan 17 02:11:26   File "/opt/conda/lib/python3.6/site-packages/torch/distributed/rpc/api.py", line 212, in shutdown
Jan 17 02:11:26     _agent.join()
Jan 17 02:11:26 RuntimeError: [/var/lib/jenkins/workspace/third_party/gloo/gloo/transport/tcp/pair.cc:572] Connection closed by peer [172.17.0.2]:51896
```

The above error suggests that allgather failed because some peer exits too early. There are two possibilities:

1. some peer leaves join() too earlier
2. some peer crashed. 

1 should not happen because, allgather guarantees that all peers have the same view on the global snapshot, so that they leave the join() function after the same amount of iterations. It's impossible (unless there is a bug in join) that one peer wants to do one more round than another peer.

So the problem is some peer crashed somewhere. Will the following case be possible?

In the `test_nested_rref_stress`, we have the following code:

```python
        for i in range(20):
            rref_of_rrefs = all_rrefs[i]
            rrefs = rref_of_rrefs.to_here()
```

Let's assume this is called on rank 0. It means that rank0 gets a fork/UserRRef from rank1's UserRRef. So, according to our protocol, rank1 will hold it's RRef alive until rank0's UserRRef is confirmed by the Owner, preventing the Owner being deleted too soon. However, it is possible that, after `_wait_all_workers()`, rank0 still didn't get the confirmation from the Owner, and `_delete_all_user_rrefs` deletes the UserRRef on rank1, which might trigger the OwnerRRef been deleted before knowing that there is a fork on rank0. 

If the above is true, then, the fix would be calling `ProcessGroupAgent::sync()` before  `_delete_all_user_rrefs()` to make sure all messages are processed before deleting UserRRefs. The `sync` API was removed, but I guess we can expose it for testing purpose? If this is indeed the problem, we can then discuss whether we want it back, or can `_wait_all_workers()` do anything about it.

a holding rrf alive for c, but been forcely erased by _delete_all_user_rrefs() which trigger owner rrefs to die

Reply:
Let's see if test pass: #32398

Reply:
This PR is for removing sync() API, but the problem this PR triggers require us to have sync() API, so this is not the right solution. Closing.